internal/homing/service.go

```diff
@@ -0,0 +1,92 @@
+package homing
+
+import (
+	"context"
+
+	"github.com/mainflux/mainflux"
+	"github.com/mainflux/mainflux/pkg/errors"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	usersObjectKey    = "users"
+	memberRelationKey = "member"
+)
+
+type Service interface {
+	Save(ctx context.Context, t Telemetry, serviceName string) error
+	GetAll(ctx context.Context, token string, pm PageMetadata) (TelemetryPage, error)
+}
+
+var _ Service = (*telemetryService)(nil)
+
+type telemetryService struct {
+	repo   TelemetryRepo
+	locSvc LocationService
+	auth   mainflux.AuthServiceClient
+}
+
+func New(repo TelemetryRepo, locSvc LocationService, auth mainflux.AuthServiceClient) Service {
+	return &telemetryService{
+		repo:   repo,
+		locSvc: locSvc,
+		auth:   auth,
+	}
+}
+
+// GetAll implements Service
+func (ts *telemetryService) GetAll(ctx context.Context, token string, pm PageMetadata) (TelemetryPage, error) {
+	res, err := ts.auth.Identify(ctx, &mainflux.Token{Value: token})
+	if err != nil {
+		return TelemetryPage{}, err
+	}
+	if err := ts.authorize(ctx, res.GetId(), usersObjectKey, memberRelationKey); err != nil {
+		return TelemetryPage{}, err
+	}
+	telemetry, err := ts.repo.RetrieveAll(ctx, pm)
+	return TelemetryPage{
+		Telemetry:    telemetry,
+		PageMetadata: pm,
+	}, err
+}
+
+// Save implements Service
+func (ts *telemetryService) Save(ctx context.Context, t Telemetry, serviceName string) error {
+	telemetry, err := ts.repo.RetrieveByIP(ctx, t.IpAddress)
```

Comment:
This effectively means that we can't save if we don't already have it saved. :)
You don't need to do this on the service level - it is not atomic and makes code more complex. Also, we'll never use explicit updates. Use the UPSERT concept (`on CONFLICT(...) DO...`). Sheets might not support this, so simply move the logic of fetching and updating on the repo layer.

Reply:
It doesn't mean we can't save if we don't have it already saved. This is not correct, we are able to create new records. I can however move the fetching and updating on the repo layer if required